src/connectors/bilibili.ts

```diff
@@ -0,0 +1,513 @@
+
+Connector.durationSelector = durationSelector;
+
+// scrobble info is shown beside the video info
+Connector.scrobbleInfoLocationSelector = '.video-info-detail';
+Connector.scrobbleInfoStyle = {
+	...Connector.scrobbleInfoStyle,
+	fontSize: '13px',
+	fontWeight: '400',
+	marginLeft: '10px',
+};
+
+Connector.isPlaying = () => {
+	const video = document.querySelector(videoSelector) as HTMLVideoElement;
+	if (!video) {
+		return false;
+	}
+	return video.currentTime > 0 && !video.paused && !video.ended;
+};
+
+// using BV number(BiliBili's video unique identifier) as unique ID
+Connector.getUniqueID = () => {
+	const currentURL = Connector.getOriginUrl();
+	const match = currentURL
+		? currentURL.match(/\/video\/(BV[0-9A-Za-z]+)/)
+		: null;
+	if (match) {
+		const bvNumber = match[1];
+		return bvNumber;
+	}
+	return null;
+};
+
+// href is like this: href="//space.bilibili.com/1459104794"
+Connector.getChannelId = () =>
+	new URL(
+		(document.querySelector(uploaderSelector) as HTMLAnchorElement)?.href ??
+			'https://bilibili.com/',
+	).pathname
+		.split('/')
+		.pop();
+
+Connector.isScrobblingAllowed = () => {
+	const tags = videoInfo.tags;
+
+	if (onlyScrobbleContainsMusicTag) {
+		if (!tags.includes('音乐')) {
+			return false;
+		}
+	} else if (useScrobbleTagFilter) {
+		if (tags.some((tag) => isIncludeElems(tag, tagFilterKeyWords))) {
+			return false;
+		}
+	}
+	return true;
+};
+
+Connector.getArtistTrack = () => {
+	getSongInfo();
+	return songInfo;
+};
+
+/*
+	helper functions	
+*/
+
+/**
+ * Check if song is changed.
+ * @returns true if new song is playing; false otherwise
+ */
+function isNewSongPlaying() {
+	const title = videoInfo.title;
+
+	if (lastVideoTitle !== title) {
+		lastVideoTitle = title;
+		return true;
+	}
+
+	return false;
+}
+
+function resetSongInfo() {
+	songInfo = null;
+}
+
+function resetVideoInfo() {
+	videoInfo = {
+		title: '',
+		upLoader: '',
+		tags: [],
+		bgmTag: null,
+	};
+}
+
+/**
+ * if the song changed, flash the songInfo object and reset state(useful when change video on the same page)
+ */
+function getSongInfo() {
+	getVideoInfo();
+	if (isNewSongPlaying()) {
+		try {
+			Connector.resetState();
+			songInfo = grepSongInfo();
+		} catch (err) {
+			const error = err as string;
+			Util.debugLog(`Error:${error}`, 'error');
+			resetSongInfo();
+			resetVideoInfo();
+		}
+	}
+}
+
+/**
+ * case insensitive filter
+ * @param items - tags to be filtered
+ * @param filterKeyWords - words that any tag contains word in this array will be removed from output
+ * @returns an array of tags does not contain any keywords defined in filterKeyWords
+ */
+function itemFilter(items: string[], filterKeyWords: string[]) {
+	return items.filter(
+		(item) =>
+			!filterKeyWords.some((keyword) => {
+				return item.toLowerCase().includes(keyword.toLowerCase());
+			}),
+	);
+}
+/**
+ * main function used to get song info.
+ * @returns \{artist, track\}
+ */
+function grepSongInfo() {
```

Comment:
Without having looked *too* thoroughly at this yet, could it be reasonable to reuse the Util.processYtVideoTitle() function here, add whatever separators and filters that won't break YT behavior that are missing to that, and then do some additional processing on top of that after?

I think this should simplify connector logic a fair bit, and will have a bonus of improving chinese track detection for all sites that use youtube's processing function.

Or is there some specific reason this is not realistic?

Reply:
Yes. It's a great function to get track and artist name from title, and I agree that it's awesome if I can adjust and reuse it, the code will be more neat and other future connector will benifit from it. Howerver, it's not suitable for this connector's logic.

I tried this func, and found it didn't work well on BiliBili as the preferences of uploaders vary in how they title their videos. thus I decide to prioritize the use of tags to enhance  accuracy. 

In this tag-centric method, the video title serves as assistance. The connector use patterns to get as much title fragment that may contain track or artist infomation as possible, push them into `possibleArtist/Track` array in pattern priority order, traverse it and find if there's any tag included in it. 

For example, consider the test link in the code (which was well-titled and tagged, though many other videos have more complicated titles) :

> videoTitle: 【官方】RADWIMPS - me me she [Music Video]
> matched possibleTrack: me me she [Music Video] 
> matched possibleArtist: 官方,RADWIMPS拉德温普斯,【官方】RADWIMPS 
> FilteredTags: me me she,RADWIMPS,摇滚
> 1. traverse possibleTrack array, found there's an tag contained in it, so take it as track name
> 2. traverse possibleArtist array, when get to the second element('RADWIMPS拉德温普斯'), found there's an tag contained in it, so the tag is going to be the artist name.

So it's hard to modify the `Util.processYtVideoTitle()` function and reused it without break it's origin behavior, since the connector needs an array containing many title fragments , not an exact string. 

I use `forEach()` method for separator array to separate the text instand of using the `Util.splitString() with the same reason. Just to ensures that, if there are multiple separators in the title, it can obtain each split result for increased accuracy.

Sorry for being this long, thanks for taking your time reading it!
Through may not being able to reuse the function, I went through the code again and tested it on more videos, then did some optimize on the connector.  If there're other inquires please let me know!

Reply:
Understood. I will have a closer look at this, and then I think we can probably merge it tomorrow 😊